177 home domain

Zagan202 opened the pull request:
# Description

Check if the issuer account has the following
- "home_domain" key value
- "home_domain"'s value is the same as the given domain where the stellar.toml will be searched for at the following location:
`https://DOMAIN/.well-known/stellar.toml`

Note: 
~I've added these tests to each SEP group test's `info.test.js` mostly because it seemed the most appropriate.~

Updates:
- I moved the tests into the `toml.test.js` tests because its more closely related to toml related requirements than /info endpoint's requirements
- ~I compare and check if every currency's issuer has the correct home_domain value.~
- I only compare and check if the active currency's issuer has the correct home_domain value.

Addresses issue: (177)

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

- [X] `npm run start:dev`
  - [X] Run SEP-6
  - [X] Run SEP-24
  - [X] Run SEP-31
  - [X] Run optional tests
  - [X] DOMAIN testanchor.stellar.org
  - [X] DOMAIN https://finclusive.com
  - [X] Run on mainnet
- [X] `DOMAIN=https://testanchor.stellar.org npx jest -i toml`
- [X] vscode debugger

Zagan202 committed:
WIP make a GET req to horizon

- Attempting to query horizon on the issuer account and see what
response we get
- From here I can check the response and compare the home
domain to the Domain

Signed-off-by: Lawrence <lawrence@stellar.org>

Zagan202 committed:
Define the horizonURL

- Change the url if its mainnet or not
- Remove unneeded imports and keypair call

Signed-off-by: Lawrence <lawrence@stellar.org>

Zagan202 committed:
compare the response's home_domain with the url found in toml

Signed-off-by: Lawrence <lawrence@stellar.org>

Zagan202 committed:
Add the home domian test to sep 24

Signed-off-by: Lawrence <lawrence@stellar.org>

Zagan202 committed:
add the home domain test to sep31

Signed-off-by: Lawrence <lawrence@stellar.org>

Zagan202 committed:
fix loggableFetch import

Signed-off-by: Lawrence <lawrence@stellar.org>

Zagan202 committed:
Port all home domain tests from info to TOML

- Its more appropriate here due to it being TOML related rather than
 /info endpoint related for each
 of the SEPs
 - This is more or less an extension of if they implemented SEP1
 properly

Signed-off-by: Lawrence <lawrence@stellar.org>

Zagan202 committed:
Refactor horizon account query

-We instead use the JS SDK to query horizon
- If the account is not there then the sdk will pass us an
error to throw.

Signed-off-by: Lawrence <lawrence@stellar.org>

Zagan202 committed:
Check every Currency's issuer's home_domain

Signed-off-by: Lawrence <lawrence@stellar.org>

JakeUrban commented on cases-SEP24/toml.test.js:
```diff
@@ -97,6 +106,18 @@ describe("TOML File", () => {
+    it("has home_domain set in the issuer account", async () => {
+      let json;
+      try {
+        for (const currency of toml.CURRENCIES) {
```
Ok so this is actually doing more than you need to do. Remember we have `process.env.CURRENCY` and `getActiveCurrency`, so when this code is run, it is only for _one currency, one issuer_.

Instead, remove the `for` loop and keep the contents. `currency` should be `process.env.CURRENCY` unless it is `null`/`undefined`, in which case you should use `getActiveCurrency`.

To break it down,
- `getActiveCurrency`
    - returns `{ enabledCurrency, infoJSON, currencies }`
    - `enabledCurrency` is the first enabled currency listed in their TOML. `currencies` are all enabled currencies. `infoJSON` as you'd probably guess is the `/info` response body.

JakeUrban replied:
Also, don't wrap the whole function in a `try`. Only `try` things you know could fail, and specifically you know what they'd fail with so you can output a good error message.

The `expect` statement for example has no reason to be in the try block. What exception would be raised as a result of that line?

JakeUrban replied:
This goes for the rest of the tests ^

JakeUrban commented on cases-SEP31/toml.test.js:
```diff
@@ -67,5 +76,17 @@ describe("TOML File", () => {
           "/",
       ).toBeTruthy();
     });
+
```
SEP-31 `getActiveCurrency` is slightly different than the SEP-6 and 24 `getActiveCurrency`, but it returns the same type. Check out the source differences between the two and understand how they're walking through the anchor's information.

JakeUrban reviewed (changes requested):
Looks really good! I can tell you took the time to see how the SDK is used and how to expect stuff. LMK if you have questions about the syntax or `async/await` stuff. 

I do have requested changes though, I think mostly due to a misunderstanding about the scope of a single test run. A test run has a few parameters that the user can provide in the UI or CLI:
- mainnet `true/false`
- protocol `sep6/sep24/sep31`
- currency `SRT/USD/etc.`
- domain `testanchor.stellar.org, etc.`
- runOptionalTests `true/false`

These are the inputs we use to determine what we test on the anchor's service during this particular test run. If the anchor wants to test multiple currencies, they do multiple test runs.

Zagan202 committed:
fix spelling

Signed-off-by: Lawrence <lawrence@stellar.org>

Zagan202 committed:
refactor test to utilize getActiveCurrency

- We're only testing the currently active issuer of the currency

Signed-off-by: Lawrence <lawrence@stellar.org>

JakeUrban commented on cases-SEP24/toml.test.js:
```diff
@@ -97,6 +107,28 @@ describe("TOML File", () => {
+          url,
+        ));
+      }
+      for (const currency of toml.CURRENCIES) {
```
This great! But I think this test will pass if `enabledCurrency` is not in `toml.CURRENCIES`, which is possible if `enabledCurrency` comes from `process.env.CURRENCY`.

I would break the for loop when you find a match but put the `expect` statement outside the loop so it's always run.

Zagan202 replied:
Good catch!
I'm changing the logic such that if `enabledCurrency` is not in `toml.CURRENCIES` we fail with a much more appropriate message `"Cannot find an issuer of the enabled currency."`
At that point if the test passes this check we'll load the account and then run the `home_domain` expect check (outside the loop).

JakeUrban replied:
that check for the passed currency should probably happen it it's own test. Can you add it? That way we can assume it's true for this test.

JakeUrban replied:
or we should just fail with from a single test with two possible error states, but thats not ideal.

Zagan202 replied:
I'll refactor these in another branch and PR in a way where each expect is in its own test.
I have a way I want it done but its slightly out of scope for this task.

JakeUrban commented on cases-SEP31/toml.test.js:
```diff
@@ -67,5 +78,25 @@ describe("TOML File", () => {
+    it("has home_domain set in the issuer account", async () => {
+      let enabledCurrency;
+      if (process.env.CURRENCY) {
+        enabledCurrency = process.env.CURRENCY;
```
Why not use `testCurrency`?

JakeUrban reviewed (commented):
A few final changes but they should be simple. After that looks good!

Zagan202 committed:
Fix bug where test will pass if loop is exhausted.

- If the forloop finishes without finding the issuer of the
enabled currency then the test will skip the expect clause.
- The logic change is first check if an issuer is found
(populated by the loop) if that fails itll report that to the user
else itll perform the check if the home domain is correct

Signed-off-by: Lawrence <lawrence@stellar.org>
